### DISCUSSION TITLE
HtmlHelp

### DISCUSSION THREAD

**Discussion by jklwn4** [nidud/asmc#2] [Q&A] [answered]
Invoking HtmlHelp requires htmlhelp.lib to be included, so adding "includelib htmlhelp.lib" to my code should solve the problem, but, while it assembles fine then, i keep getting this when linking (MS´s linker):

HTMLHELP.LIB(htmlhelp.obj) : error LNK2019: unresolved external symbol __imp__RegQueryValueExA@24 referenced in function _LoadHHCtrl@0
HTMLHELP.LIB(htmlhelp.obj) : error LNK2019: unresolved external symbol __imp__RegOpenKeyExA@20 referenced in function _LoadHHCtrl@0
HTMLHELP.LIB(htmlhelp.obj) : error LNK2019: unresolved external symbol __imp__RegCloseKey@4 referenced in function _LoadHHCtrl@0  

providing protos for these three APIs doesn´t help either - what can i do?

**Comment by nidud** [comment]
> Invoking HtmlHelp requires htmlhelp.lib

This should be added to the header file, so I will look into that.

The DLL file for HtmlHelp is [hhctrl.ocx](https://learn.microsoft.com/en-us/windows/win32/api/htmlhelp/nf-htmlhelp-htmlhelpa) so to test the import directly using:

`asmc -pe test.asm`

```
include stdio.inc
include tchar.inc

option dllimport:<hhctrl.ocx>
HtmlHelpA proto WINAPI :ptr, :ptr, :UINT, :DWORD

.code

_tmain proc

    .if !printf("Htmlhelp\n")

        HtmlHelpA(NULL, NULL, 0, 0)
    .endif
    .return(0)

_tmain endp

    end _tstart
```

If this works there's probably something wrong with the .LIB file you using or some additional `includelib` issues for the missing imports.

**Comment by nidud** [comment]
I added an update to Asmc to handle the extension for the DLL file. The pragma options for this (added to the header file) is:

`.pragma comment(lib, htmlhelp, hhctrl.ocx)
`

The tools used for creating import libraries from system DLL's depends on a .DLL extension so creating the HtmlHelp library is a bit tricky. I modified the [dlldef](https://github.com/nidud/asmc/tree/master/source/tools/dlldef) tool a bit to add argument count for 32-bit imports.

Note that importing DLL's differ based on bit-ness, so you have to build the tool in the bit-ness as the library (I assume 32-bit in your case). The default for the `makefile` is 64-bit, so the command should be:

```
asmc -pe dlldef.asm
dlldef hhctrl.ocx
```

This produce a .DEF file like this:
```
LIBRARY hhctrl.ocx
EXPORTS
AuthorMsg@16
DllCanUnloadNow@12
DllGetClassObject@12
DllRegisterServer@12
DllUnregisterServer@12
HhWindowThread@4
HtmlHelpA@16
HtmlHelpW@0
LoadHHA@8
doWinMain@8
```

Note that this is based on searching the binary object for the firs occurrence of a RET instruction and use the return size if found, so the result may not be correct. The header file only define two of these functions and the `HtmlHelpW@0` should be `HtmlHelpW@16`. As for the other functions I have no idea if they are correct or not.

After correcting this I tried building the library using LIB:

`lib /out:htmlhelp.lib /machine:X32 /def:hhctrl.def
`

This linked but failed to execute. Converting the .DEF to LIBW input worked for some reason.

```
++_AuthorMsg@16.'hhctrl.ocx'
++_DllCanUnloadNow@12.'hhctrl.ocx'
++_DllGetClassObject@12.'hhctrl.ocx'
++_DllRegisterServer@12.'hhctrl.ocx'
++_DllUnregisterServer@12.'hhctrl.ocx'
++_HhWindowThread@4.'hhctrl.ocx'
++_HtmlHelpA@16.'hhctrl.ocx'
++_HtmlHelpW@16.'hhctrl.ocx'
++_LoadHHA@8.'hhctrl.ocx'
++_doWinMain@8.'hhctrl.ocx'

libw /n /c /q /b /fac /ii htmlhelp.lib @hhctrl.def
```

**Comment by jklwn4** [comment]
Many thanks, this indeed creates a working htmlhelp.lib in 32 bit! 

Now i have the same problem in 64 bit. Libw as far as i understand it cannot create 64 bit .lib files, or am i wrong here?

**Comment by nidud** [comment, marked as answer]
It's simpler in 64-bit (and __cdecl in 32-bit) as you can use the export names directly. The installer [uses LIBW](https://github.com/nidud/asmc/blob/1d998ff1b48f31e92509f17f556858f9cf1ad5f5/lib/x64/makefile#L204) to create the import libraries for 64-bit by default, but optionally also LIB. 

The .DEF file for LIBW:

```
++AuthorMsg.'hhctrl.ocx'
++HhWindowThread.'hhctrl.ocx'
++HtmlHelpA.'hhctrl.ocx'
++HtmlHelpW.'hhctrl.ocx'
++LoadHHA.'hhctrl.ocx'
++doWinMain.'hhctrl.ocx'
```
Command:
```
libw /n /c /q /b /fac /i6 htmlhelp.lib @hhctrl.def
```

The .DEF file for LIB:

```
LIBRARY hhctrl.ocx
EXPORTS
AuthorMsg
HhWindowThread
HtmlHelpA
HtmlHelpW
LoadHHA
doWinMain
```
Command:
```
lib /out:htmlhelp.lib /machine:X64 /def:hhctrl.def
```

Note that the Dll* functions was removed here:
```
hhctrl.def : warning LNK4104: export of symbol 'DllCanUnloadNow' should be PRIVATE
hhctrl.def : warning LNK4104: export of symbol 'DllGetClassObject' should be PRIVATE
hhctrl.def : warning LNK4104: export of symbol 'DllRegisterServer' should be PRIVATE
hhctrl.def : warning LNK4104: export of symbol 'DllUnregisterServer' should be PRIVATE
```

So the output from `dlldef` needs some manual editing (or improvement). It also adds "quotes" to the names that needs to be removed in 64-bit.

**Comment by jklwn4** [comment]
Linkw doesn´t work for me in 64 bit, but Lib does - thanks!

In the meantime i found this

`@echo off

REM Usage: dll2lib [32|64] some-file.dll
REM
REM Generates some-file.lib from some-file.dll, making an intermediate
REM some-file.def from the results of dumpbin /exports some-file.dll.
REM
REM Requires 'dumpbin' and 'lib' in PATH - run from VS developer prompt.
REM 
REM Script inspired by http://stackoverflow.com/questions/9946322/how-to-generate-an-import-library-lib-file-from-a-dll
SETLOCAL
if "%1"=="32" (set machine=x86) else (set machine=x64)
set dll_file=%2
set dll_file_no_ext=%~n2
set exports_file=%dll_file_no_ext%-exports.txt
set def_file=%dll_file_no_ext%.def
set lib_file=%dll_file_no_ext%.lib
set lib_name=%dll_file_no_ext%

dumpbin /exports %dll_file% > %exports_file%

echo LIBRARY %lib_name% > %def_file%
echo EXPORTS >> %def_file%
for /f "skip=19 tokens=1,4" %%A in (%exports_file%) do if NOT "%%B" == "" (echo %%B @%%A >> %def_file%)

lib /def:%def_file% /out:%lib_file% /machine:%machine%

REM Clean up temporary intermediate files
del %exports_file% %def_file% %dll_file_no_ext%.exp`

on stackoverflow, which works fine as well.
